### PR TITLE
Fix doxygen warnings in MQTT Library documentation

### PR DESCRIPTION
**Pull request by sarenameas**
- Fix doxygen warnings throughout mqtt documentation.
- Add documentation to mqtt_lightweight.c static functions where missing (this was a warning)
- Add all words to the lexicon.txt
- Opportunistically Fix other code warnings found in the unit test and mqtt.c
- Ignore private functions in mqtt_state.h and .c

Doxygen configs and other infrastructure is to be in a separate PR.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: libraries/standard/mqtt/include/mqtt.h [hunk 1]
```diff
@@ -64,9 +63,9 @@ typedef uint32_t (* MQTTGetCurrentTimeFunc_t )( void );
  * @param[in] pPacketInfo Information on the type of incoming MQTT packet.
  * @param[in] pDeserializedInfo Deserialized information from incoming packet.
  */
-typedef void (* MQTTEventCallback_t )( MQTTContext_t * pContext,
-                                       MQTTPacketInfo_t * pPacketInfo,
-                                       MQTTDeserializedInfo_t * pDeserializedInfo );
+typedef void (* MQTTEventCallback_t )( struct MQTTContext * pContext,
```
Comment (sarenameas) [inline libraries/standard/mqtt/include/mqtt.h:66]: There are no misra issues or compiler warnings with this change.

File: libraries/standard/mqtt/include/mqtt_lightweight.h [hunk 2]
```diff
@@ -55,27 +67,11 @@
 #define MQTT_PACKET_TYPE_PINGRESP       ( ( uint8_t ) 0xD0U )  /**< @brief PINGRESP (server-to-client). */
 #define MQTT_PACKET_TYPE_DISCONNECT     ( ( uint8_t ) 0xE0U )  /**< @brief DISCONNECT (client-to-server). */
 
-
 /**
  * @brief The size of MQTT PUBACK, PUBREC, PUBREL, and PUBCOMP packets, per MQTT spec.
  */
 #define MQTT_PUBLISH_ACK_PACKET_SIZE    ( 4UL )
 
-struct MQTTFixedBuffer;
```
Comment (sarenameas) [inline libraries/standard/mqtt/include/mqtt_lightweight.h:64]: Doxygen does not easily support forward declarations like this, so i revert to our original style of `typedef struct mystrcyt {} mystruct_t;`

File: libraries/standard/mqtt/src/mqtt_lightweight.c [hunk 3]
```diff
@@ -291,6 +297,133 @@ static uint8_t * encodeString( uint8_t * pDestination,
 static size_t getRemainingLength( TransportRecv_t recvFunc,
                                   NetworkContext_t * pNetworkContext );
 
+/**
```
Comment (sarenameas) [inline libraries/standard/mqtt/src/mqtt_lightweight.c:300]: Most of this is copied from the v4_beta docs with updates for new parameters and return code names.

File: libraries/standard/mqtt/include/mqtt.h [hunk 4]
```diff
@@ -35,14 +39,9 @@
  */
 #define MQTT_PACKET_ID_INVALID    ( ( uint16_t ) 0U )
 
-struct MQTTPubAckInfo;
-typedef struct MQTTPubAckInfo         MQTTPubAckInfo_t;
-
+/* Forward declared structs for the MQTTEventCallback_t. */
```
Comment (muneebahmed10) [inline libraries/standard/mqtt/include/mqtt.h:42]: If we're able to forward declare the structs but not the typedefs, why not declare all of the structs here, instead of just the ones in the callback? It would let us easily see at the top of the file what structs are used

File: libraries/standard/mqtt/src/mqtt_state.c [hunk 5]
```diff
@@ -615,6 +621,10 @@ static uint16_t stateSelect( const MQTTContext_t * pMqttContext,
 
 /*-----------------------------------------------------------*/
 
+/**
+ * @cond DOXYGEN_IGNORE
```
[the remark this answers is not in the document]
Comment (sarenameas) [reply libraries/standard/mqtt/src/mqtt_state.c:625]: Yes

File: libraries/standard/mqtt/include/mqtt.h (hunk 4 above)
Comment (sarenameas) [reply libraries/standard/mqtt/include/mqtt.h:42]: Because the context struct needs the definition of the MQTTEventCallback_t

File: libraries/standard/mqtt/src/mqtt_lightweight.c [hunk 6]
```diff
@@ -291,6 +297,133 @@ static uint8_t * encodeString( uint8_t * pDestination,
...
+                                                 size_t qos0Minimum );
+
+/**
+ * @brief Process incoming publish flags.
+ *
+ * @param[in] publishFlags Incoming publish flags.
+ * @param[in, out] pPublishInfo Pointer to #MQTTPublishInfo_t struct where
+ * output will be written.
+ *
+ * @return #MQTTSuccess, #MQTTBadResponse.
```
[the remark this answers is not in the document]
Comment (sarenameas) [reply libraries/standard/mqtt/src/mqtt_lightweight.c:333]: Sure thanks

File: libraries/standard/mqtt/src/mqtt_lightweight.c [hunk 7]
```diff
@@ -291,6 +297,133 @@ static uint8_t * encodeString( uint8_t * pDestination,
...
+ *
+ * @param[in] packetType The packet type to check.
+ *
+ * @return `true` if the packet type is valid; `false` otherwise.
+ */
+static bool incomingPacketValid( uint8_t packetType );
+
+/**
+ * @brief Check the remaining length of an incoming PUBLISH packet against some
+ * value for QoS 0 or QoS 1 and 2.
```
[the remark this answers is not in the document]
Comment (sarenameas) [reply libraries/standard/mqtt/src/mqtt_lightweight.c:311]: sure

File: libraries/standard/mqtt/src/mqtt_lightweight.c [hunk 8]
```diff
@@ -291,6 +297,133 @@ static uint8_t * encodeString( uint8_t * pDestination,
...
+ * @return #MQTTSuccess or #MQTTBadResponse.
+ */
+static MQTTStatus_t checkPublishRemainingLength( size_t remainingLength,
+                                                 MQTTQoS_t qos,
+                                                 size_t qos0Minimum );
+
+/**
+ * @brief Process incoming publish flags.
+ *
+ * @param[in] publishFlags Incoming publish flags.
```
[the remark this answers is not in the document]
Comment (sarenameas) [reply libraries/standard/mqtt/src/mqtt_lightweight.c:329]: Sure

File: libraries/standard/mqtt/include/mqtt.h (hunk 4 above)
Comment (sarenameas) [reply libraries/standard/mqtt/include/mqtt.h:42]: Adding the third struct per misread confusion.

File: libraries/standard/mqtt/src/mqtt_state.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (sarenameas) [reply libraries/standard/mqtt/src/mqtt_state.c:625]: It warns function does not have documentation. I thought a bit more and I am going to add documentation in here since we don't want to the function to not appear at all.

File: libraries/standard/mqtt/include/mqtt_lightweight.h [hunk 9]
```diff
@@ -19,12 +19,23 @@
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
  */
 
+/**
+ * @file mqtt_lightweight.h
+ * @brief User-facing functions for serializing MQTT 3.1.1 packets. This header
+ * should be included for building a light-weight MQTT client bypassing the
+ * stateful CSDK MQTT library API in mqtt.h.
```
Comment (muneebahmed10) [inline libraries/standard/mqtt/include/mqtt_lightweight.h:26]: Would it be accurate to call this stateful? I know we keep state records for QoS 1/2 publishes, but we don't keep track of acks for un/subscribes. We could just use "managed"
```suggestion
 * managed CSDK MQTT library API in mqtt.h.
```

File: libraries/standard/mqtt/include/mqtt_lightweight.h [hunk 10]
```diff
@@ -19,12 +19,23 @@
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
  */
 
+/**
+ * @file mqtt_lightweight.h
+ * @brief User-facing functions for serializing MQTT 3.1.1 packets. This header
+ * should be included for building a light-weight MQTT client bypassing the
```
Comment (muneebahmed10) [inline libraries/standard/mqtt/include/mqtt_lightweight.h:25]: Nit: consistency with other usage
```suggestion
 * should be included for building a lightweight MQTT client bypassing the
```

File: libraries/standard/mqtt/src/mqtt_lightweight.c [hunk 11]
```diff
@@ -19,6 +19,10 @@
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
  */
 
+/**
+ * @file mqtt_lightweight.c
+ * @brief Implements the user facing functions in mqtt_lightweight.h.
```
Comment (muneebahmed10) [inline libraries/standard/mqtt/src/mqtt_lightweight.c:24]: Nit: consistency with header file
```suggestion
 * @brief Implements the user-facing functions in mqtt_lightweight.h.
```

File: libraries/standard/mqtt/include/mqtt.h [hunk 12]
```diff
@@ -64,9 +64,9 @@ typedef uint32_t (* MQTTGetCurrentTimeFunc_t )( void );
  * @param[in] pPacketInfo Information on the type of incoming MQTT packet.
  * @param[in] pDeserializedInfo Deserialized information from incoming packet.
  */
-typedef void (* MQTTEventCallback_t )( MQTTContext_t * pContext,
-                                       MQTTPacketInfo_t * pPacketInfo,
-                                       MQTTDeserializedInfo_t * pDeserializedInfo );
+typedef void (* MQTTEventCallback_t )( struct MQTTContext * pContext,
+                                       struct MQTTPacketInfo * pPacketInfo,
```
Comment (leegeth) [inline libraries/standard/mqtt/include/mqtt.h:68]: Would it be better to move the structure definitions along with their type definitions above these callbacks so that we can consistently use the same type everywhere?

File: libraries/standard/mqtt/src/mqtt_lightweight.c [hunk 13]
```diff
@@ -291,6 +297,133 @@ static uint8_t * encodeString( uint8_t * pDestination,
...
+ *
+ * Converts the packet from a stream of bytes to an #MQTTPublishInfo_t and
+ * extracts the packet identifier. Also prints out debug log messages about the
+ * packet.
+ *
+ * @param[in] pIncomingPacket Pointer to an MQTT packet struct representing a
+ * PUBLISH.
+ * @param[out] pPacketId Packet identifier of the PUBLISH.
+ * @param[out] pPublishInfo Pointer to #MQTTPublishInfo_t where write output is
+ * written.
```
Comment (muneebahmed10) [inline libraries/standard/mqtt/src/mqtt_lightweight.c:391]:
```suggestion
 * @param[out] pPublishInfo Pointer to #MQTTPublishInfo_t where output is
 * written.
```

File: libraries/standard/mqtt/src/mqtt_lightweight.c [hunk 14]
```diff
@@ -291,6 +297,133 @@ static uint8_t * encodeString( uint8_t * pDestination,
...
+ *
+ * @return #MQTTSuccess if PUBLISH is valid; #MQTTBadResponse
+ * if the PUBLISH packet doesn't follow MQTT spec.
+ */
+static MQTTStatus_t deserializePublish( const MQTTPacketInfo_t * pIncomingPacket,
+                                        uint16_t * pPacketId,
+                                        MQTTPublishInfo_t * pPublishInfo );
+
+/**
+ * @brief Deserialize a UNSUBACK, PUBACK, PUBREC, PUBREL, or PUBCOMP packet.
```
Comment (muneebahmed10) [inline libraries/standard/mqtt/src/mqtt_lightweight.c:401]:
```suggestion
 * @brief Deserialize an UNSUBACK, PUBACK, PUBREC, PUBREL, or PUBCOMP packet.
```

File: libraries/standard/mqtt/include/mqtt_lightweight.h [hunk 15]
```diff
@@ -19,12 +19,23 @@
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
  */
 
+/**
+ * @file mqtt_lightweight.h
+ * @brief User-facing functions for serializing MQTT 3.1.1 packets. This header
```
Comment (leegeth) [inline libraries/standard/mqtt/include/mqtt_lightweight.h:24]: serializing and deserializing

File: libraries/standard/mqtt/include/mqtt.h (hunk 12 above)
Comment (sarenameas) [reply libraries/standard/mqtt/include/mqtt.h:68]: Then what about MQTTEventCallback_t definition in MQTTContext_t

File: libraries/standard/mqtt/include/mqtt_lightweight.h (hunk 9 above)
Comment (sarenameas) [reply libraries/standard/mqtt/include/mqtt_lightweight.h:26]: Sure

File: libraries/standard/mqtt/include/mqtt.h (hunk 12 above)
Comment (leegeth) [reply libraries/standard/mqtt/include/mqtt.h:68]: oh I see. That is an interesting problem. What is the doxygen issue if we forward declare the structure and typedef it?

Comment (sarenameas) [reply libraries/standard/mqtt/include/mqtt.h:68]: Then we cannot make references to the members of the struct in the documentation like `#MQTTContext_t.networkBuffer` .

File: libraries/standard/mqtt/src/mqtt_lightweight.c (hunk 11 above)
Comment (sarenameas) [reply libraries/standard/mqtt/src/mqtt_lightweight.c:24]: sure

File: libraries/standard/mqtt/include/mqtt_lightweight.h (hunk 10 above)
Comment (sarenameas) [reply libraries/standard/mqtt/include/mqtt_lightweight.h:25]: sure

File: libraries/standard/mqtt/src/mqtt_lightweight.c (hunk 13 above)
Comment (sarenameas) [reply libraries/standard/mqtt/src/mqtt_lightweight.c:391]: thanks

File: libraries/standard/mqtt/src/mqtt_lightweight.c (hunk 14 above)
Comment (sarenameas) [reply libraries/standard/mqtt/src/mqtt_lightweight.c:401]: thanks

File: libraries/standard/mqtt/include/mqtt_lightweight.h (hunk 15 above)
Comment (sarenameas) [reply libraries/standard/mqtt/include/mqtt_lightweight.h:24]: thanks

File: libraries/standard/mqtt/include/mqtt.h (hunk 12 above)
Comment (muneebahmed10) [reply libraries/standard/mqtt/include/mqtt.h:68]: You might have tried this already, but I'm wondering if a `@struct` tag helps at all

Comment (sarenameas) [reply libraries/standard/mqtt/include/mqtt.h:68]: Yea tried

File: libraries/standard/mqtt/src/mqtt_state.c [hunk 16]
```diff
@@ -1096,5 +1176,6 @@ const char * MQTT_State_strerror( MQTTPublishState_t state )
 
     return str;
 }
+/** @endcond */
```
Comment (muneebahmed10) [inline libraries/standard/mqtt/src/mqtt_state.c:1179]: I see a lot of `@endcond` in this file. Does there not need to be a corresponding `@cond`?

File: libraries/standard/mqtt/include/mqtt_state.h [hunk 17]
```diff
@@ -19,29 +19,44 @@
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
  */
 
+/**
+ * @file mqtt_state.h
+ * @brief User-facing functions for keeping state of MQTT 3.1.1 packets.
```
Comment (muneebahmed10) [inline libraries/standard/mqtt/include/mqtt_state.h:24]: Have we agreed that all of these functions are meant to be user-facing? Also I'd rephrase to something like "keeping state of MQTT publish deliveries", since it's the publishes we're keeping state of, not the packets themselves.

File: libraries/standard/mqtt/src/mqtt_state.c (hunk 16 above)
Comment (sarenameas) [reply libraries/standard/mqtt/src/mqtt_state.c:1179]: I need to remove those thanks

File: libraries/standard/mqtt/include/mqtt_state.h (hunk 17 above)
Comment (sarenameas) [reply libraries/standard/mqtt/include/mqtt_state.h:24]: No only the MQTT_PublishToResend it public. I will reword this thanks.

File: libraries/standard/mqtt/src/mqtt_state.c [hunk 18]
```diff
@@ -19,6 +19,10 @@
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
  */
 
+/**
+ * @file mqtt_state.c
+ * @brief Implements the user-facing functions in mqtt_state.h.
```
Comment (muneebahmed10) [inline libraries/standard/mqtt/src/mqtt_state.c:24]: I don't think all the state functions are user-facing

Comment (sarenameas) [reply libraries/standard/mqtt/src/mqtt_state.c:24]: Ah right this needs to be update too thanks
